rust/src/http2/range.rs

```diff
@@ -38,52 +40,45 @@ pub struct HTTPContentRange {
 }
 
 pub fn http2_parse_content_range_star<'a>(input: &'a [u8]) -> IResult<&'a [u8], HTTPContentRange> {
-    let (i2, _) = char!(input, '*')?;
-    let (i2, _) = char!(i2, '/')?;
-    let (i2, size) = map_res!(i2, map_res!(digit1, std::str::from_utf8), i64::from_str)?;
+    let (i2, _) = char('*')(input)?;
+    let (i2, _) = char('/')(i2)?;
+    let (i2, size) = map_res(map_res(digit1, std::str::from_utf8), i64::from_str)(i2)?;
     return Ok((
         i2,
         HTTPContentRange {
             start: -1,
             end: -1,
-            size: size,
+            size,
         },
     ));
 }
 
 pub fn http2_parse_content_range_def<'a>(input: &'a [u8]) -> IResult<&'a [u8], HTTPContentRange> {
-    let (i2, start) = map_res!(input, map_res!(digit1, std::str::from_utf8), i64::from_str)?;
-    let (i2, _) = char!(i2, '-')?;
-    let (i2, end) = map_res!(i2, map_res!(digit1, std::str::from_utf8), i64::from_str)?;
-    let (i2, _) = char!(i2, '/')?;
-    let (i2, size) = alt!(
-        i2,
-        value!(-1, char!('*')) | map_res!(map_res!(digit1, std::str::from_utf8), i64::from_str)
-    )?;
-    return Ok((
-        i2,
-        HTTPContentRange {
-            start: start,
-            end: end,
-            size: size,
-        },
-    ));
+    let (i2, start) = map_res(map_res(digit1, std::str::from_utf8), i64::from_str)(input)?;
+    let (i2, _) = char('-')(i2)?;
+    let (i2, end) = map_res(map_res(digit1, std::str::from_utf8), i64::from_str)(i2)?;
+    let (i2, _) = char('/')(i2)?;
+    let (i2, size) = alt((
+        value(-1, char('*')),
+        map_res(map_res(digit1, std::str::from_utf8), i64::from_str),
+    ))(i2)?;
+    return Ok((i2, HTTPContentRange { start, end, size }));
 }
 
 fn http2_parse_content_range<'a>(input: &'a [u8]) -> IResult<&'a [u8], HTTPContentRange> {
-    let (i2, _) = take_while!(input, |c| c == b' ')?;
-    let (i2, _) = take_till!(i2, |c| c == b' ')?;
-    let (i2, _) = take_while!(i2, |c| c == b' ')?;
-    return alt!(
-        i2,
-        http2_parse_content_range_star | http2_parse_content_range_def
-    );
+    let (i2, _) = take_while(|c| c == b' ')(input)?;
+    let (i2, _) = take_till(|c| c == b' ')(i2)?;
+    let (i2, _) = take_while(|c| c == b' ')(i2)?;
+    return alt((
+        http2_parse_content_range_star,
+        http2_parse_content_range_def,
+    ))(i2);
 }
 
 pub fn http2_parse_check_content_range<'a>(input: &'a [u8]) -> IResult<&'a [u8], HTTPContentRange> {
     let (rem, v) = http2_parse_content_range(input)?;
     if v.start > v.end {
-        return Err(Err::Error((rem, ErrorKind::Verify)));
+        return Err(Err::Error(make_error(rem, ErrorKind::Verify)));
```

Comment:
Do we need the 3 `Err` `Err::Error` and `make_error` ?

Reply:
That part of nom errors also irritates me ...
- The first `Err` is the `Result` variant, so required
- `Err::Error` is the nom error variant (could be `Error`, or `Failure` to stop without possible recovery), it is needed
- `make_error` is a convenience function to create a generic nom error. This is not required but only useful for genericity and to ease future code updates. It can be removed it to return only the tuple.

Reply:
Well, I let you make the PR to nom to get something better ;-)

Should we macro it meanwhile ?

Reply:
Indeed, we can shorten the notation using a macro.
Do you want me to push commits for this?

Reply:
If you feel it is idiomatic

Reply:
I've got used to this pattern the more I've worked with nom7.  I'm not sure a macro would help much.

Reply:
It's up to you! Personally, I am now used to this syntax and agree with Jason, so I do not think a macro would really help. In fact, I find it confusing since it tends to hide types and makes resolving compiler errors a bit more difficult.